File: unit-test-practice-3/zip/characterization-tests/src/test/csharp/dojo.Tests/ParserCharacterizationTest.cs
using NUnit.Framework;
using System;

namespace dojo.Tests
{
    [TestFixture]
    public class ParserCharacterizationTest
    {
        private Parser parser = new Parser();

        [Test]
        public void Test_Good_Address()
        {
            Test("2040 Budaörs Valami utca 5.", "2040", "Budaörs", "Valami utca 5.");
            Test("Budaörs 2040 Valami utca 5.", "2040", "Budaörs", "Valami utca 5.");
            Test("Valami utca 5. Budaörs 2040", "2040", "Budaörs", "Valami utca 5.");
            Test("Valami utca 5. 2040 Budaörs", "2040", "Budaörs", "Valami utca 5.");
        }

        [Test]
        public void Test_Bad_Address()
        {
            Test("Valami 2040 Budaörs utca 5.", "2040", "Budaörs", "Valami utca 5.");
            Test("Valami 2040 utca 5. Budaörs", "2040", "Budaörs", "Valami utca 5.");
            Test("Valami Budaörs utca 5. 2040", "2040", "Budaörs", "Valami utca 5.");
            Test("Budaörs Valami 2040 utca 5.", "2040", "Budaörs", "Valami utca 5.");
            Test("2040 Valami Budaörs utca 5.", "2040", "Budaörs", "Valami utca 5.");
        }

        [Test]
        public void Test_Duplicated_Parts()
        {
            Test("2040 Budaörs Valami utca 2040", "2040", "Budaörs", "Valami utca 2040");
            Test("2040 Budaörs 2040 Valami utca", "2040", "Budaörs", "2040 Valami utca");
            Test("Budaörs 2040 Budaörs utca 5.", "2040", "Budaörs", "Budaörs utca 5.");
            Test("Budaörs Budaörs utca 5. 2040", "2040", "Budaörs", "Budaörs utca 5.");
            Test("Budaörs Kis Budaörs utca 5. 2040", "2040", "Budaörs", "Kis Budaörs utca 5.");
        }

        [Test]
        public void Test_Numbers()
        {
            Test("123 1234", "1234", null, "123");
            Test("1234 123", "1234", null, "123");
        }

        [Test]
        public void Test_Address_With_Whitespace()
        {
            Test(" 2040", "2040", null, null);
            Test("\t2040", "2040", null, null);
            Test("\n2040", "2040", null, null);
            Test("\r2040", "2040", null, null);
            Test("2040 ", "2040", null, null);
            Test("2040\t", "2040", null, null);
            Test("2040\n", "2040", null, null);
            Test("2040\r", "2040", null, null);
        }

        private void Test(string input, params string[] expectedResult)
        {
            string[] result = parser.Parse(input);
            Assert.That(result, Is.EqualTo(expectedResult));
        }
    }
}

File: unit-test-practice-3/zip/characterization-tests/src/test/java/dojo/ParserCharacterizationTest.java
package dojo;

import static org.assertj.core.api.Assertions.assertThat;

import org.testng.annotations.Test;

public class ParserCharacterizationTest {

	private Parser parser = new Parser();
	
	@Test
	public void test_good_address() throws Exception {
		test("2040 Budaörs Valami utca 5.", "2040", "Budaörs", "Valami utca 5.");
		test("Budaörs 2040 Valami utca 5.", "2040", "Budaörs", "Valami utca 5.");
		test("Valami utca 5. Budaörs 2040", "2040", "Budaörs", "Valami utca 5.");
		test("Valami utca 5. 2040 Budaörs", "2040", "Budaörs", "Valami utca 5.");
	}
	
	@Test
	public void test_bad_address() throws Exception {
		test("Valami 2040 Budaörs utca 5.", "2040", "Budaörs", "Valami utca 5.");
		test("Valami 2040 utca 5. Budaörs", "2040", "Budaörs", "Valami utca 5.");
		test("Valami Budaörs utca 5. 2040", "2040", "Budaörs", "Valami utca 5.");
		test("Budaörs Valami 2040 utca 5.", "2040", "Budaörs", "Valami utca 5.");
		test("2040 Valami Budaörs utca 5.", "2040", "Budaörs", "Valami utca 5.");
	}
	
	@Test
	public void test_duplicated_parts() throws Exception {
		test("2040 Budaörs Valami utca 2040", "2040", "Budaörs", "Valami utca 2040");
		test("2040 Budaörs 2040 Valami utca", "2040", "Budaörs", "2040 Valami utca");
		test("Budaörs 2040 Budaörs utca 5.", "2040", "Budaörs", "Budaörs utca 5.");
		test("Budaörs Budaörs utca 5. 2040", "2040", "Budaörs", "Budaörs utca 5.");
		test("Budaörs Kis Budaörs utca 5. 2040", "2040", "Budaörs", "Kis Budaörs utca 5.");
	}
	
	@Test
	public void test_numbers() throws Exception {
		test("123 1234", "1234", null, "123");
		test("1234 123", "1234", null, "123");
	}
	
	@Test
	public void test_address_with_whitespace() throws Exception {
		test(" 2040", "2040", null, null);
		test("\t2040", "2040", null, null);
		test("\n2040", "2040", null, null);
		test("\r2040", "2040", null, null);
		test("2040 ", "2040", null, null);
		test("2040\t", "2040", null, null);
		test("2040\n", "2040", null, null);
		test("2040\r", "2040", null, null);
	}
	
	private void test(String input, String ... expectedResult) {
		String[] result = parser.parse(input);
		assertThat(result).containsExactly(expectedResult);
	}
	
}
